samples/settings/src/settings.c

```diff
@@ -0,0 +1,185 @@
+				"{\"status\": \"failed\", "
+				"\"msg\": \"failed to save setting %s:%s\"}\n", name, val);
+		} else {
+			shell_fprintf(shell, SHELL_ERROR,
+				      "Failed to save setting %s:%s\n", name, val);
+		}
+
+		return -ENOEXEC;
+	}
+
+	if (json_output) {
+		shell_fprintf(shell, SHELL_VT100_COLOR_GREEN,
+			"{\"status\": \"success\", \"msg\": \"setting %s saved as %s\"}\n",
+			name, val);
+	} else {
+		shell_fprintf(shell, SHELL_NORMAL,
+			      "Setting %s saved as %s\n", name, val);
+	}
+
+	return 0;
+}
+
+struct settings_read_callback_params {
+	const struct shell *shell_ptr;
+	bool value_found;
+};
+
+static int settings_read_callback(const char *key,
+				  size_t len,
+				  settings_read_cb read_cb,
+				  void            *cb_arg,
+				  void            *param)
+{
+	ssize_t num_read_bytes = MIN(len, SETTINGS_MAX_VAL_LEN);
+	uint8_t buffer[num_read_bytes + 1];
+	struct settings_read_callback_params *params = param;
+
+	/* Process only the exact match and ignore descendants of the searched name */
+	if (settings_name_next(key, NULL) != 0) {
+		return 0;
+	}
+
+	params->value_found = true;
+	num_read_bytes = read_cb(cb_arg, buffer, num_read_bytes);
+
+	if (num_read_bytes < 0) {
+		if (json_output) {
+			shell_fprintf(params->shell_ptr, SHELL_VT100_COLOR_RED,
+				"{\"status\": \"failed\", "
+				"\"msg\": \"failed to read value: %d\"}\n",
+				(int)num_read_bytes);
+		} else {
+			shell_error(params->shell_ptr, "Failed to read value: %d",
+				(int) num_read_bytes);
+		}
+
+		return 0;
+	}
+
+	/*  add NULL to the last position in the buffer */
+	buffer[num_read_bytes] = 0x00;
+
+	if (json_output) {
+		shell_fprintf(params->shell_ptr, SHELL_VT100_COLOR_GREEN,
+			"{\"status\": \"success\", \"value\": \"%s\"}\n", buffer);
+	} else {
+		shell_fprintf(params->shell_ptr, SHELL_VT100_COLOR_GREEN, "%s\n", buffer);
+	}
+
+	if (len > SETTINGS_MAX_VAL_LEN) {
+		shell_print(params->shell_ptr, "(The output has been truncated)");
+	}
+
+	return 0;
+}
+
+static int cmd_settings_get(const struct shell *shell, size_t argc,
+			    char *argv[])
+{
+	int err;
+	const char *name = argv[1];
+
+	json_output = false;
+	if (argc >= 3) {
+		if (strcmp(argv[2], "--json") == 0) {
+			json_output = true;
+		}
+	}
+
+	struct settings_read_callback_params params = {
+		.shell_ptr = shell,
+		.value_found = false
+	};
+
+	err = settings_load_subtree_direct(name, settings_read_callback, &params);
+
+	if (err) {
+		if (json_output) {
+			shell_fprintf(shell, SHELL_VT100_COLOR_RED,
+				"{\"status\": \"failed\", "
+				"\"msg\": \"failed to load settings: %d\"}\n", err);
+		} else {
+			shell_error(shell, "Failed to load settings: %d", err);
+		}
+	} else if (!params.value_found) {
+		if (json_output) {
+			shell_fprintf(shell, SHELL_VT100_COLOR_RED,
+				"{\"status\": \"failed\", \"msg\": \"setting not found\"}\n");
+		} else {
+			shell_error(shell, "Setting not found");
+		}
+	}
+
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(settings_commands,
```

Comment:
Or even better improve the string provided in `SHELL_CMD_ARG()` macros. Unfortunately Zephyr shell commands are not consistent with how the help message should look like, but it is worth searching for a best candidate, so that users can learn how to use shell without leaving shell and readin comments that implement those shell commands.

Reply:
> Unfortunately Zephyr shell commands are not consistent with how the help message should look like, but it is worth searching for a best candidate, so that users can learn how to use shell without leaving shell and readin comments that implement those shell commands.

@mniestroj - I'm not sure I understand the full context, but do you think this is the sort of thing that should be fixed in this PR, or is it okay to split this off into a separate backlog item?

Reply:
We can split this off and improve in next PR.

Reply:
Anyway, I improved the help string to help users understand without leaving the shell

```
uart:~$ settings
settings - Settings commands
Subcommands:
  set  :set a module config
        usage:
        $ settings set <key> <value>
        example:
        $ settings set golioth/psk pre-shared-key-value
        $ seetings set golioth/psk-id psk-identity-value
  get  :get a module config value
        usage:
        $ settings get <key>
        example:
        $ settings get golioth/psk
        $ settings get golioth/psk-id

```

Reply:
what do you think?